File: src/java/magmac/app/compile/node/Node.java
package magmac.app.compile.node;

import magmac.api.Option;
import magmac.api.Tuple2;
import magmac.api.collect.list.List;
import magmac.api.iter.Iter;
import magmac.app.compile.error.CompileResult;
import magmac.app.lang.Serializable;

import java.util.function.Function;

public interface Node {
    CompileResult<NodeList> findNodeListOrError(String key);

    CompileResult<Node> findNodeOrError(String key);

    Iter<Tuple2<String, Node>> iterNodes();

    String display();

    String format(int depth);

    Iter<Tuple2<String, NodeList>> iterNodeLists();

    Node withString(String key, String value);

    Option<String> findString(String key);

    Node merge(Node other);

    boolean is(String type);

    Node retype(String type);

    Node withNodeList(String key, NodeList values);

    Option<NodeList> findNodeList(String key);

    Node withNode(String key, Node value);

    Option<Node> findNode(String key);

    Iter<Tuple2<String, String>> iterStrings();

    boolean hasNodeList(String key);

    CompileResult<Tuple2<Node, NodeList>> removeNodeListOrError(String key);

    Option<Tuple2<Node, NodeList>> removeNodeList(String key);

    boolean hasContent();

    CompileResult<Tuple2<Node, String>> removeString(String key);

    CompileResult<Tuple2<Node, Node>> removeNodeOrError(String key);

    <T> Node withNodeListAndSerializer(String key, List<T> list, Function<T, Node> serializer);

    <T> Node withNodeAndSerializer(String key, T element, Function<T, Node> serializer);

    default <T extends Serializable> Node withNodeListSerialized(String key, List<T> list) {
        return this.withNodeListAndSerializer(key, list, Serializable::serialize);
    }

    default <T extends Serializable> Node withNodeSerialized(String key, T element) {
        return this.withNodeAndSerializer(key, element, Serializable::serialize);
    }

    Option<Tuple2<Node, Node>> removeNode(String key);
}

File: src/web/magmac/app/compile/node/Node.ts
import { Option } from "../../../../magmac/api/Option";
import { Tuple2 } from "../../../../magmac/api/Tuple2";
import { List } from "../../../../magmac/api/collect/list/List";
import { Iter } from "../../../../magmac/api/iter/Iter";
import { CompileResult } from "../../../../magmac/app/compile/error/CompileResult";
import { Serializable } from "../../../../magmac/app/lang/Serializable";
import { Function } from "../../../../java/util/function/Function";
export interface Node {
	findNodeListOrError(key : String) : CompileResult<NodeList>;
	findNodeOrError(key : String) : CompileResult<Node>;
	iterNodes() : Iter<Tuple2<String, Node>>;
	display() : String;
	format(depth : int) : String;
	iterNodeLists() : Iter<Tuple2<String, NodeList>>;
	withString(key : String, value : String) : Node;
	findString(key : String) : Option<String>;
	merge(other : Node) : Node;
	is(type : String) : boolean;
	retype(type : String) : Node;
	withNodeList(key : String, values : NodeList) : Node;
	findNodeList(key : String) : Option<NodeList>;
	withNode(key : String, value : Node) : Node;
	findNode(key : String) : Option<Node>;
	iterStrings() : Iter<Tuple2<String, String>>;
	hasNodeList(key : String) : boolean;
	removeNodeListOrError(key : String) : CompileResult<Tuple2<Node, NodeList>>;
	removeNodeList(key : String) : Option<Tuple2<Node, NodeList>>;
	isEmpty() : boolean;
	removeString(key : String) : CompileResult<Tuple2<Node, String>>;
	removeNodeOrError(key : String) : CompileResult<Tuple2<Node, Node>>;
	withNodeListAndSerializer(key : String, list : List<T>, serializer : Function<T, Node>) : Node;
	withNodeAndSerializer(key : String, element : T, serializer : Function<T, Node>) : Node;
	withNodeListSerialized(key : String, list : List<T>) : Node {return 0.withNodeListAndSerializer( 0, 0, 0.serialize);;}
	withNodeSerialized(key : String, element : T) : Node {return 0.withNodeAndSerializer( 0, 0, 0.serialize);;}
	removeNode(key : String) : Option<Tuple2<Node, Node>>;
}
